src/pool/srv_pool.c

```diff
@@ -1300,98 +1333,221 @@ ds_pool_enable_exclude(void)
 	pool_disable_exclude = false;
 }
 
+static int
+alloc_event_set(struct pool_svc_event_set **event_set)
+{
+	D_ALLOC_PTR(*event_set);
+	if (*event_set == NULL)
+		return -DER_NOMEM;
+	return 0;
+}
+
+static void
+free_event_set(struct pool_svc_event_set **event_set)
+{
+	D_FREE((*event_set)->pss_buf);
+	D_FREE(*event_set);
+}
+
+static int
+add_to_event_set(struct pool_svc_event_set *event_set, d_rank_t rank, uint64_t incarnation,
+		 enum crt_event_source src, enum crt_event_type type)
+{
+	int i;
+
+	/* Find rank in event_set. */
+	for (i = 0; i < event_set->pss_len; i++)
+		if (event_set->pss_buf[i].psv_rank == rank)
+			break;
+
+	/* If not found, prepare to add a new event. */
+	if (i == event_set->pss_len) {
+		if (event_set->pss_len == event_set->pss_cap) {
+			uint32_t               cap;
+			struct pool_svc_event *buf;
+
+			if (event_set->pss_cap == 0)
+				cap = 1;
+			else
+				cap = 2 * event_set->pss_cap;
+			D_REALLOC_ARRAY(buf, event_set->pss_buf, event_set->pss_cap, cap);
+			if (buf == NULL)
+				return -DER_NOMEM;
+			event_set->pss_buf = buf;
+			event_set->pss_cap = cap;
+		}
+		event_set->pss_len++;
+	}
+
+	event_set->pss_buf[i].psv_rank        = rank;
```

Comment:
Do we need compare incarnation if in existed case?

Reply:
We do not, I think. The `crt` module should have ordered the events properly (`swim` compares incarnations before updating the SWIM status of a rank). Also, some `GRPMOD` events have `0` incarnations:

```
09/26-01:31:55.86 [...] queue_event() 7db09892: queuing event: rank=6 inc=1885680475287322624 src=1 type=1
09/26-01:31:56.22 [...] queue_event() 7db09892: queuing event: rank=6 inc=0 src=2 type=1
```

And the `0`-incarnation event really occurred after, and should override, the one with a proper incarnation.